### PR TITLE
#296 preserve tag's whitespaces for alpanumeric ordering index

### LINKED ISSUE #296: Preserve whitespaces for alphabetic Tag ordering
default

BTW You have to create `AlterIndexSQL` class

### PR DESCRIPTION
**Pull request by artemiy312**
Closes #296

### CODE REVIEW

File: catalog/models.py [hunk 1]
```diff
@@ -230,7 +230,7 @@ class TagQuerySet(models.QuerySet):
     def order_by_alphanumeric(self):
         """Sort the Tag by name's alphabetic chars and then by numeric chars."""
         return self.annotate(
-            tag_name=Substring(models.F('name'), models.Value('[a-zA-Zа-яА-Я]+')),
+            tag_name=Substring(models.F('name'), models.Value('[a-zA-Zа-яА-Я\s]+')),
```
Comment (duker33) [inline catalog/models.py]: add these symbols too: `-_,:;`

File: catalog/models_operations.py [hunk 2]
```diff
@@ -1,57 +1,134 @@
...
+
+    @abc.abstractmethod
+    def execute(self, table: str, schema_editor: BaseDatabaseSchemaEditor):
+        """Execute SQL operation."""
+
+
+class AddIndex(IndexSQL):
+
+    def __init__(self, name: str, columns: typing.List[str]):
+        self.name = name
```
Comment (duker33) [inline catalog/models_operations.py]: i'll be more good to call super instead: `super().__init__(name)`. Up to you

File: catalog/models_operations.py [hunk 3]
```diff
@@ -1,57 +1,134 @@
...
+    reversible = True
+
+    def __init__(self, operation: IndexOperation):
+        self.operation = operation
+
+    def state_forwards(self, app_label, state):
+        """We have to implement this method for Operation interface."""
+
+    def database_forwards(self, app_label, schema_editor, from_state, to_state):
+        self.operation.database_backwards(app_label, schema_editor, from_state, to_state)
```
Comment (duker33) [inline catalog/models_operations.py]:
```suggestion
        self.operation.database_forwards(app_label, schema_editor, from_state, to_state)
```

File: catalog/models_operations.py [hunk 4]
```diff
@@ -1,57 +1,134 @@
...
+
+    def database_backwards(self, app_label, schema_editor, from_state, to_state):
+        self.operation.database_forwards(app_label, schema_editor, from_state, to_state)
```
Comment (duker33) [inline catalog/models_operations.py]:
```suggestion
        self.operation.database_backwards(app_label, schema_editor, from_state, to_state)
```

File: catalog/models_operations.py [hunk 5]
```diff
@@ -1,57 +1,134 @@
...
-                f'DROP INDEX {self._index_name(table_name)};'
-            )
+            self.backward.execute(table_name, schema_editor)
 
     def describe(self):
-        return f'Create index {self.name} for {self.model_name}'
+        return f'Operate the index {self.name} for {self.model_name}'
+
+
+class RevertOperation(Operation):
```
Comment (duker33) [inline catalog/models_operations.py]: Let's not use verbs in class names:
```suggestion
class ReversedOperation(Operation):
```

Or even better `Reversed`. We are already in `operations` module

File: catalog/models_operations.py (hunk 3 above)
Comment (artemiy312) [reply catalog/models_operations.py]: The implementation is correct. It allows to revert wrapped operation

**Pull request merged**